Solver: Cache Concretization Results

johnwparent opened the pull request:
# Spack Concretization Cache
This PR introduces an initial attempt to cache concretization results. I'm sure the design/function can be improved, this is PR is intended to be an initial implementation to be built from.

This PR:

* Adds two config options, one to provide a location for the concretization cache on disk (default is `~/.spack/concretization`), the other assigns a limit to how many cache entries we allow (default is 1k)
* Adds the concretization cache to `spack clean` as a standalone clean operation as well as part of `spack clean -a`
* Will prune the cache during each solver run to ensure the cache does not grow unchecked
    - Cache entries are pruned based on age of the entry, this is subject to change, input encouraged.
    - Cache entries are allowed to accumulate to max concretizer cache entries, and once that limit is reached, the oldest 10% are pruned automatically.

* Makes a number of changes to `SpackSolverSetup` and related classes to ensure output of solver setup is deterministic
    - Adds a test to validate this

### Caching approach:

Cache hashes are the sha256 hash of the string representation of:

* All `.lp` asp files involved in a given concretization run
* The output of `SpackSolverSetup.setup` split by newline, sorted, and rejoined into a single string

This hash determines if a cache hit has occurred. 
Cache entries are stored as a file named by the given hash under a directory with a name based on the first two characters of the cache hash (similar to the source cache).

The contents of a cache entry are `spack.solver.asp.Result` objects serialized to disk via json, when cache hits are encountered, these result json representations are read from disk and recomposed into `spack.solver.asp.Result` objects, allowing the reproduction of a solver run exactly as if it had just occurred with no loss of context or statistics. 


## Benchmarks

3 specs were selected for bench-marking, `readline`, `hdf5+mpi build_system=cmake` and `paraview+mpi+python`.
These packages were selected as they represent a spectrum of DAG sizes, and range from simple to including virtuals, and splicing.
Benchmarks were performed on an M1 Mac

### `readline`

Without a cache:
```
    setup          3.518s
    cache-check     0.041s
    ordering       0.034s
    load           0.648s
    ground         2.362s
    solve          1.174s
    construct_specs     0.329s
    total          8.093s
```
With a cache:
```
    setup          3.459s
    cache-check     0.041s
    ordering       0.034s
    total          3.510s
```

### `hdf5+mpi build_system=cmake`
Without a cache:
```
    setup          3.523s
    cache-check     0.041s
    ordering       0.035s
    load           0.642s
    ground         2.401s
    solve          1.510s
    construct_specs     0.426s
    total          8.557s
```
With a cache:
```
    setup          3.523s
    cache-check     0.063s
    ordering       0.034s
    total          3.595s
```
### `paraview+mpi+python`
Without a cache:
```
    setup          7.983s
    cache-check     0.091s
    ordering       0.076s
    load           1.473s
    ground         5.570s
    solve          5.335s
    construct_specs     1.104s
    total         21.566s
```
With a cache:
```
    setup         7.985ms
    cache-check  0.121s
    ordering     0.0898s
    total      8.196s
```

johnwparent committed:
in-situ

progress

Major updates

Overhaul

johnwparent committed:
more deterministic (but its not enough)

johnwparent committed:
Add clean operation for concretizer cache

rm cruft

johnwparent committed:
Error assurance

johnwparent committed:
Ensure deterministic output

write out asp problem no ordering

Style improvements

johnwparent committed:
Test: deterministic asp output ordering

johnwparent commented:
cc @tgamblin @becker33

alalazo commented on lib/spack/spack/solver/asp.py:
```diff
@@ -542,6 +556,224 @@ def format_unsolved(unsolved_specs):
+        return result
+
+
+class ConcretizationCache:
```
In view of #45189, can we add new code related to caches to a separate file, e.g.
```
spack/solver/caches.py
```
so to minimize conflicts?

johnwparent replied:
Reorganized! I think there will be some unavoidable conflicts but this should reduce them at least.

johnwparent replied:
@alalazo After resolving the circular import indicated by the CI failure, re-organizing this into its own module resulted in either a significant re-write of this module (the diff is _much_ larger) or another circular import that fails the CI check. I think conflicts with this should be pretty minimal as the vast majority of the code is a single block, and the rest is ordering sets, which should be fairly evident when looking at a diff conflict.

If you still want this in another module, I can PR the re-org separately (no changes, just moving code) so we can both rebase on that to reduce conflicts

kwryankrattiger replied:
As unfortunate as it is to have yet another semi-large class in `asp.py`, I think holding off and doing a proper refactor later is probably fine for now. We want to be able to use the _before_ the compilers as node changes if possible.

johnwparent commented:
Addressing the import check failure will require move around some additional blocks of code into new modules, pushing up a pass at that now, but very open to alternative suggestions

haampie commented:
> concretization cache should be left out of the spack.caches module

Go for it

haampie commented on lib/spack/spack/solver/cache.py:
```diff
@@ -0,0 +1,163 @@
+    def remove(self, problem: str) -> bool:
+        """Removes cache entry associated with problem"""
+        cache_path = self._cache_path(problem)
+        if cache_path.exists():
```
for example here remove the branch, run unlink and catch the error instead

haampie commented on lib/spack/spack/solver/cache.py:
```diff
@@ -0,0 +1,163 @@
+                "config:concretization_cache", spack.paths.default_conc_cache_path
+            )
+        self.root = pathlib.Path(spack.util.path.canonicalize_path(root))
+        self.cleanup()
```
I wouldn't run this in the constructor, but only on `store`

johnwparent replied:
Agreed constructor is wrong place, just put it there because I couldn't decide where it should actually go. You think store is also right? I think both places are called to often...

haampie commented on lib/spack/spack/solver/cache.py:
```diff
@@ -0,0 +1,163 @@
+        # TODO: determine a better default
+        if self.root.exists():
+            entry_limit = spack.config.get("config:concretization_cache_limit", 1000)
+            entries = [x for x in self.cache_entries()]
```
I think it's both brittle and very inefficient to traverse the file system tree here. I certainly don't want Spack to do another 1000 stat calls on top of the current 7K for package.py files.

Instead, keep track of the total number of files added and bytes written when `store` is called. That means you need a write lock for a file with totals.

johnwparent replied:
Inefficient for sure, I just wanted _something_ here as a cleanup operation so we could iterate on it in review. This is also terrible because it's run every time the conc cache is constructed, which should not go in as is, would love a suggestion for where/when we should be running this. 

Will refactor approach. There's some other metadata that would be a useful store, and I'd also like a pruning policy based on whether or not a given concretization is even possible anymore given changes to packages/lp files over time.

Will also make process safe.

johnwparent commented:
> I only quickly glanced at this, but what I've seen does not look process safe.

You're right, will address, thanks.

johnwparent committed:
Add manifest for cleanup and process saftey

johnwparent committed:
Ensure abstract specs are not cast to dict

johnwparent committed:
Refactor to use convenience method

johnwparent committed:
Update config scheme

johnwparent committed:
Add text fixture and config option to diable conc cache

johnwparent committed:
bug fixes and determinism

Style

more determinism (its still not enough)

Repair tests

style

johnwparent commented:
Looks like this still needs to handle splicing a bit better, will investigate further monday

johnwparent committed:
replace os.rename w/ shutil.move

Enables better cross device handling

johnwparent committed:
Make concretization cache part of misc cache

johnwparent committed:
reframe testing strategy

Style cleanup

Refactor use of test fixtures

alalazo commented on lib/spack/spack/schema/config.py:
```diff
@@ -71,6 +71,10 @@
             "license_dir": {"type": "string"},
             "source_cache": {"type": "string"},
             "misc_cache": {"type": "string"},
+            "enable_concretization_cache": {"type": "boolean"},
+            "concretization_cache_path": {"type": "string"},
+            "concretization_cache_entry_limit": {"type": "integer", "minimum": 1},
+            "concretization_cache_byte_limit": {"type": "integer", "minimum": 800},
```
Why not using a sub-object in the configuration? It would be:
```yaml
concretization_cache:
    enable: true
    path: ....
    entry_limit: ...   
    size_limit: ...
```
compared to:
```yaml
enable_concretization_cache: ...
concretization_cache_path: ....
concretization_cache_entry_limit: ...   
concretization_cache_byte_limit: ...
```
I would also use the generic `size` instead of `byte`, cause that wouldn't need to change if later we allow `5GB` or something similar as a valid value.

johnwparent replied:
> Why not using a sub-object in the configuration? It would be:

Initially I wasn't intending to add so many options, but yeah, that's way better, great suggestion. Thanks

> cause that wouldn't need to change if later we allow 5GB or something similar as a valid value.

Is your thought the user would indicate something like `5GB` as a string in their config? I'm hesitant to add extra parsing and validation for config values, but it would be nice to not have to just give the config a giant number.

alalazo replied:
> Is your thought the user would indicate something like 5GB as a string in their config? I'm hesitant to add extra parsing and validation for config values, but it would be nice to not have to just give the config a giant number.

Yes, what I mean is that we don't have to implement it in this PR, but later we can say "if it's an int, it's supposed to be bytes, otherwise we can use properly formatted strings". If we call the attribute `byte_limit` we are committed to understand the entry as bytes.

EDIT: Even though, being very pedantic, `byte_limit: 5G` works too :laughing:

johnwparent replied:
> being very pedantic, byte_limit: 5G works too 😆

I don't think it's being pedantic, any file size limit is inherently dealing in bytes.

I do agree though that a generic "size" is a better name.

alalazo replied:
Yeah, sorry, the `pedantic` part was me thinking that if we removed the `B` from:
```
size_limit: 5GB
```
`byte_limits` works too, cause the unit of measure is already in the attribute name :slightly_smiling_face:

johnwparent replied:
> Yeah, sorry, the pedantic part was me thinking that if we removed the B from:

No need for apology, I was/am your side (I'm also a fan of being pendantic from time to time) 😄 
That's actually really funny.


> byte_limits works too, cause the unit of measure is already in the attribute name 🙂

I honestly don't have an opinion, I'll go with whatever you/the general consensus prefers.

johnwparent committed:
repair concretization cache metadata read issues

johnwparent committed:
refactor conc cache config

johnwparent committed:
refactor file iteration approach

remove literal typing

johnwparent committed:
restore behavior accidentally removed in refactor

wip

johnwparent committed:
Add conc cache test

style

Fixup new test

johnwparent commented:
Gitlab failures are currently rouge pipeline failures.

johnwparent committed:
Add some more docstrings

johnwparent committed:
refactor cache parsing error handling

johnwparent committed:
Add checks to conc cache test to ensure we're actually using it

style

kwryankrattiger commented on lib/spack/spack/schema/config.py:
```diff
@@ -58,6 +58,15 @@
+                "type": "object",
+                "properties": {
+                    "enable": {"type": "boolean"},
+                    "path": {"type": "string"},
```
Could we do `url` instead of `path`. We will eventually want to point this at remote locations and I am hoping to just reuse the connection schema from `.../schema/mirrors.py`

johnwparent replied:
NP!

johnwparent committed:
Refactor filecache

Convert filecache to:
* use pathlib
* take ranges
* dynamic timeouts per transaction

Missed filecache change

smaller filecache fixups

Finalize filecache refactor

Rm all non pathlib changes to filecache

johnwparent committed:
Refactor conc cache

johnwparent committed:
Add manifest dump and cleanup operation calls

johnwparent committed:
RM cruft

johnwparent committed:
Rename schema

johnwparent committed:
Post rebase refactor

johnwparent committed:
mypy == sad

johnwparent committed:
Further style check sadness

kwryankrattiger commented on lib/spack/spack/schema/config.py:
```diff
@@ -58,6 +58,15 @@
                     {"type": "string"},  # deprecated
                 ]
             },
+            "concretization_cache": {
```
This still needs an entry in `lib/spack/docs`

kwryankrattiger replied:
Things to note:

pruning policy
fs layout of cache

johnwparent committed:
maintain access check model

kwryankrattiger commented on lib/spack/spack/solver/asp.py:
```diff
@@ -538,6 +558,364 @@ def format_unsolved(unsolved_specs):
+            )
+        self.root = pathlib.Path(spack.util.path.canonicalize_path(root))
+        self._fc = FileCache(self.root)
+        self._cache_manifest = self.root / ".cache_manifest"
```
`_cache_manifest` should not include `self.root`, that is/shoudl be added by `FileCache.cache_path`.

kwryankrattiger commented on lib/spack/spack/solver/asp.py:
```diff
@@ -538,6 +558,364 @@ def format_unsolved(unsolved_specs):
+        """Prunes the concretization cache according to configured size and entry
+        count limits. Cleanup is done in FIFO ordering."""
+        # TODO: determine a better default
+        entry_limit = spack.config.get("config:concretization_cache:entry_limit", 1000)
```
Maybe there could be overrides for these to say "do not prune dynamically, I want to manage this myself"

-1 or 0 are both somewhat common options for this.

johnwparent committed:
Add docs

johnwparent committed:
bug fix and improved limit handling

johnwparent committed:
Update schema for better limit handling

kwryankrattiger commented on lib/spack/docs/config_yaml.rst:
```diff
@@ -334,3 +336,52 @@ create a new alias called ``inst`` that will always call ``install -v``:
+``concretization_cache:enable``
+-------------------------------
+
+When set to ``true``, Spack will create a cache for the solver output of
```
```suggestion
When set to ``true``, Spack will utilize a cache of solver outputs from
```

kwryankrattiger commented on lib/spack/docs/config_yaml.rst:
```diff
@@ -334,3 +336,52 @@ create a new alias called ``inst`` that will always call ``install -v``:
+
+When set to ``true``, Spack will create a cache for the solver output of
+successful concretization runs. When enabled, Spack will check the concretization
+cache prior to running the solver, and if a previous request to solve a given
```
```suggestion
cache prior to running the solver. If a previous request to solve a given
```

johnwparent committed:
LoL

kwryankrattiger commented on lib/spack/docs/config_yaml.rst:
```diff
@@ -334,3 +336,52 @@ create a new alias called ``inst`` that will always call ``install -v``:
+after each concretization run; if Spack has stored more results than the limit allows, the
+oldest concretization results are pruned until 10% of the limit has been removed.
+
+Setting this value to 0 or -1 disables the automatic pruning. It is expected users will be
```
Let's pick just one value. I think I prefer `0`, and we can set the range in the schema to be `"minimum": 0,`.

kwryankrattiger commented on lib/spack/docs/config_yaml.rst:
```diff
@@ -334,3 +336,52 @@ create a new alias called ``inst`` that will always call ``install -v``:
+after each concretization run; if Spack has stored more results than the limit allows, the
+oldest concretization results are pruned until 10% of the limit has been removed.
+
+Setting this value to 0 or -1 disables the automatic pruning. It is expected users will be
```
Same here, let's just use `0` and reflect that in the schema

johnwparent committed:
Refactor based on doc review

alalazo commented:
Did anybody try any performance benchmark on this PR? Otherwise I can try to compare with develop on a cold and hot cache.

kwryankrattiger commented:
@johnwparent had shared some numbers previously.

I agree it would be nice to have a record of that in the PR too.

kwryankrattiger reviewed (approved):
Approving now, will merge after performance numbers are posted.

johnwparent commented:
@kwryankrattiger @alalazo benchmarks have been added to the PR description

alalazo commented:
Do we have a lockfile per cache entry:
```console
$ ls -larth ~/.spack/cache/concretization/12
total 1,1M
-rwxrwxr-x 1 culpo culpo    0 feb 28 18:36 .128f2dc2834e4e295aca1f50be39c5003340cbcaaff2db8e94ee565d92c3bea1.lock
drwxrwxr-x 3 culpo culpo 4,0K feb 28 18:36 ..
-rw-rw-r-- 1 culpo culpo 1,1M feb 28 18:36 128f2dc2834e4e295aca1f50be39c5003340cbcaaff2db8e94ee565d92c3bea1
drwxrwxr-x 2 culpo culpo 4,0K feb 28 18:36 .
```
?

alalazo commented:
I guess we can probably optimize in a later PR to use a single lock per the entire cache, like we do for store and stages (and lock different bytes base on the entry).

haampie commented:
@kwryankrattiger why did you merge this? Did you try it out? Please check with other reviewers before merging.

I would suggest to disable this feature, because getting cache hits reliably for even the simplest spec is impossible:

```
concretization $ spack -d  spec zlib |& grep -E '(hit|miss)'
==> Concretization cache miss at 18/18607911a1944d36e1a49f3728783e0a957a0a21c390138f22044afc1715ba4e
concretization $ spack -d  spec zlib |& grep -E '(hit|miss)'
==> Concretization cache miss at 28/286c42d84b58dbbc59d6e9197fb31560b1fafe6d1044a69b04f546e4f6e02f6d
concretization $ spack -d  spec zlib |& grep -E '(hit|miss)'
==> Concretization cache hit at 28/286c42d84b58dbbc59d6e9197fb31560b1fafe6d1044a69b04f546e4f6e02f6d
concretization $ spack -d  spec zlib |& grep -E '(hit|miss)'
==> Concretization cache hit at 18/18607911a1944d36e1a49f3728783e0a957a0a21c390138f22044afc1715ba4e
concretization $ spack -d  spec zlib |& grep -E '(hit|miss)'
==> Concretization cache miss at de/de4b86777b02dbba18b370a56ce6d40213ab03d775432dd888909622b9687cff
concretization $ spack -d  spec zlib |& grep -E '(hit|miss)'
==> Concretization cache miss at dc/dcdd9c18c0480908ae5676b4ec026a091f5c999e4d1f1bb366c0dd4e990f3f79
concretization $ spack -d  spec zlib |& grep -E '(hit|miss)'
==> Concretization cache miss at 4f/4f998c4b067724eab7de84ebcc43ba2bb3450648819d75fe0a14b816e7d24a0e
concretization $ spack -d  spec zlib |& grep -E '(hit|miss)'
==> Concretization cache miss at 43/4301e1911df4c87eaab53d3991c22986c8d64948ce6129564aaed2cfad570028

2 hits, 6 misses. That's really bad.

kwryankrattiger commented:
I had tried it out and we had discussed this PR in another meeting with @tgamblin and @alalazo and it was fine to merge.

The issues causing non determinism were found post merge. The issues are known and being resolved by @johnwparent now. It isn't breaking workflows so no one jumped to revert.

haampie commented:
> We don't put mirror misc cache configs in mirrors.yaml for example.

Not the best example given that `misc_cache` is a combination of repo metadata, compiler output and build cache indices.

I still think it's better to put it in `concretizer` given it has various sub-options.

But in any case, it would have been good if I were given the chance to look at this before merge given that I had started review initially.
